### PR TITLE
chore: sync template to template-repo@061d9ab (template-2026.04)

### PR DESCRIPTION
**Pull request by agorokh**
## Summary

Propagates the canonical template from `76e62d2` (pinned Apr 4) to `061d9abf` (today), including the root-cause fix for the hook drift that's been silently aborting `issue-driven-coding-orchestrator` runs in this repo (template-repo #91/#92). 52 template commits worth of improvements land in one PR.

## Why now

The orchestrator stopped reaching `git push` / `gh pr create` on EPIC #86 (second occurrence on this machine). Root cause upstream: `PostToolUse:Bash` was a `type: "prompt"` hook expecting the small model to emit literal `PASS` — any prose drift halted continuation after every successful bash. Fixed in template-repo PR #92 (replaced with deterministic command hooks + test contract preventing recurrence). This PR lands that fix here plus all other template evolution since Apr 4.

### Hook delivery (the unblock)
- `.claude/settings.base.json` + regenerated `.claude/settings.json` now use **command hooks** for flow control.
- `scripts/hook_protect_main.sh` + `scripts/hook_protect_main_impl.py` — blocks git commit/push on protected branches.
- `scripts/hook_sensitive_file_guard.sh` — blocks edits to `.env`, lockfiles, private keys, `archive/`, `secrets/`.
- `scripts/hook_bash_pre_tool.sh` — unified bash pre-tool hook.
- `tests/test_hook_scripts.py` — 97/101 tests pass locally. The 4 failures are Windows-only path-parsing edge cases in `hook_protect_main_impl.py` (pre-existing template behavior around `git -C <win-path>`; upstream follow-up, non-blocking).
- `docs/00_Core/HOOK_DESIGN.md` — convention doc: command hooks for flow control, prompts for advisory only.

### New Claude Code + Cursor skills
- `orchestrate`, `resolve-pr`, `post-merge`, `dependency-review`, `learner`, `ci-check`, `new-project-setup`, `release-notes`, `github-issue-creator`.

### Infrastructure
- Self-learning scaffolding: `.claude/rules/learned/{domain,local,universal}` + cursor mirror.
- Post-merge deterministic steward: `.claude/agents/post-merge-steward.md`, `.github/workflows/vault-automerge.yml`, `post-merge-notify.yml`, `scripts/post_merge_sync.sh` split into sync/vault phases.
- PR analysis: `.github/workflows/codeql.yml`, `pr-pain-detection.yml`, `qodo-review.yml`, `.pr_agent.toml`, `.gemini/`, `.sourcery.yaml`, `.secrets.baseline`.
- Docs: `docs/00_Core/CHILD_REPOS.md`, `PROPAGATION_STRATEGY.md`. Removed stale `CODEX.md`, `WARP.md`.
- New tests: bootstrap_knowledge, fleet_inventory_refresh, vault_audit, session_debrief_ingest.

### Hand-merged conflicts (project-specific content preserved)
- `pyproject.toml`: unioned template security floor (`detect-secrets`, `pyyaml`, `pygments>=2.20.0`) with project deps (`websockets`, `lupa`). Kept `coaching` extras alongside new `training`.
- `Makefile`: unioned `ci-conventional` / `ci-secrets` / `init-knowledge` targets with project-specific `ci-csp-api` / `ci-csp-ui-safety`.
- `README.md`: kept project-specific Requirements/Installation/License (MIT, AC-specific flow).

### Bump pin
- `.copier-answers.yml` `_commit` manually bumped to `061d9abf0018ee5589bb2304d5aa5841815c82df`. The template has no git tags, so `copier update` doesn't auto-bump the pin — manual in the absence of `template-YYYY.MM` tags.

## Test plan

- [x] `py -3 scripts/merge_settings.py --no-local` — regenerated settings.json parses, 2 advisory prompt hooks remain (LOAD + SQL DDL).
- [x] `pytest tests/test_hook_scripts.py` — 97/101 passing (4 Windows-specific failures pre-existing upstream).
- [ ] CI green on this branch (GitHub Actions).
- [ ] `issue-driven-coding-orchestrator` should now reach `git push` / `gh pr create` without stalling (to be confirmed on the next run — e.g. continuing #86).

## Out of scope (follow-ups)

- [ ] Resolve 4 Windows-only hook impl failures in `hook_protect_main_impl.py` → file upstream issue on template-repo.
- [ ] Continue EPIC #86 work (rig screen Phase-2 scaffold was stashed before this sync; rebase `feat/issue-86-rig-screen-phase2-launcher-and-apps` onto `main` once this lands).
- [ ] Propose tagging `template-2026.04` on template-repo so future `copier update` auto-bumps `_commit`.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: .claude/agents/post-merge-steward.md [hunk 1]
```diff
@@ -22,35 +22,86 @@ memory: project
...
 ### Phase B — classification
 
-3. Run `python3 scripts/post_merge_classify.py --pr <P>`.
+3. Run `python3 scripts/post_merge_classify.py --pr <P>` and read stdout. The merged PR may also have a bot comment from `.github/workflows/post-merge-notify.yml`.
+
+### Phase C — vault SAVE (LLM judgment, then deterministic ship)
+
+4. **SAVE** per `docs/00_Core/SESSION_LIFECYCLE.md`, editing **only** files under `docs/01_Vault/`:
+   - Update `docs/01_Vault/ProjectTemplate/00_System/Next Session Handoff.md`: move merged work to “What was delivered” (PR link, merge SHA, date); refresh “Resume here” / “What remains” (include Phase B follow-ups).
+   - Update `docs/01_Vault/ProjectTemplate/00_System/Current Focus.md` if this PR closed the active focus.
```
[the remark this answers is not in the document]
Comment (agorokh) [reply .claude/agents/post-merge-steward.md]: Fixed in 2e4943c. The canonical `ProjectTemplate` paths on lines 42-43 were rewritten to `docs/01_Vault/AcCopilotTrainer/...` — the template sync came in as a merge PR rather than via `copier update`, so `scripts/copier_post_copy.py` (`_rewrite_tree`) never ran on this file. Thanks for the catch.

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -5,12 +5,31 @@ repos:
...
         args: ["--maxkb=512"]
+      - id: no-commit-to-branch
+        args: ["--branch", "main", "--branch", "master"]
 
   - repo: https://github.com/astral-sh/ruff-pre-commit
     rev: v0.8.6
     hooks:
       - id: ruff
         args: [--fix]
       - id: ruff-format
```
[the remark this answers is not in the document]
Comment (agorokh) [reply .pre-commit-config.yaml:21]: Fixed in 2e4943c: bumped `ruff-pre-commit` from `v0.8.6` to `v0.15.12` to match `pyproject.toml` (`ruff>=0.15.11`) so pre-commit and `make ci-fast` run the same ruff and agree on lint/format.

File: tools/process_miner/github_client.py [hunk 3]
```diff
@@ -231,3 +233,90 @@ def get_linked_issues(
...
+            raise
+        if not isinstance(data, dict):
+            return None
+        if data.get("type") != "file":
+            return None
+        b64 = data.get("content")
+        if not isinstance(b64, str):
+            return None
+        raw = base64.b64decode(b64.replace("\n", ""))
+        return raw.decode("utf-8", errors="replace")
```
[the remark this answers is not in the document]
Comment (agorokh) [reply tools/process_miner/github_client.py:330]: Fixed in 2e4943c: guarded `base64.b64decode` in `get_contents_text` with a try/except for `binascii.Error` / `ValueError`, returning `None` on non-decodable payloads. Preserves the "text or None" contract.

File: .claude/pitfalls-hub.json [hunk 4]
```diff
@@ -0,0 +1,6 @@
+{
+  "hub_repo": "agorokh/template-repo",
+  "hub_path": "docs/01_Vault/ProjectTemplate/pitfalls",
```
[the remark this answers is not in the document]
Comment (agorokh) [reply .claude/pitfalls-hub.json:3]: Declining this one — the `hub_path` staying as `docs/01_Vault/ProjectTemplate/pitfalls` is intentional. `.claude/pitfalls-hub.json` is in `SKIP_REWRITE_FILES` in `scripts/copier_post_copy.py` (line 102) with the explicit comment: `hub_path intentionally references template-repo (hub-spoke)`. The hub lives at `agorokh/template-repo` and this client repo points at its `ProjectTemplate/pitfalls` directory; if we rewrote it the skill would fetch from a non-existent path in the hub repo.

File: .github/workflows/vault-automerge.yml [hunk 5]
```diff
@@ -1,64 +1,74 @@
...
+# applies — if required reviews/checks block the bot, a human must merge.
+#
+# This decouples the post-merge-steward agent from `git push origin main`:
+# the agent opens a labeled PR, this workflow guards scope and ships it.
+
 on:
   pull_request:
-    branches: [main]
-    types: [opened, synchronize, reopened, labeled, unlabeled]
+    types: [opened, synchronize, reopened, labeled]
```
[the remark this answers is not in the document]
Comment (agorokh) [reply .github/workflows/vault-automerge.yml:12]: Valid observation, but deferring to an upstream `template-repo` issue rather than changing it in this sync PR. The workflow comes in verbatim from `template-repo@061d9ab` and reverting the unlabeled-trigger removal here would cause template drift. Will file the follow-up against `agorokh/template-repo` — fix belongs one level up so every child repo inherits it.

File: .github/workflows/vault-automerge.yml [hunk 6]
```diff
@@ -1,64 +1,74 @@
...
-      - name: Disable auto-merge when vault-only label is removed
-        env:
-          GH_TOKEN: ${{ github.token }}
-          PR_NUMBER: ${{ github.event.pull_request.number }}
-          REPO: ${{ github.repository }}
-        run: gh pr merge "$PR_NUMBER" --repo "$REPO" --disable-auto || true
-
   guard-and-automerge:
-    if: contains(github.event.pull_request.labels.*.name, 'vault-only') && github.event.pull_request.head.repo.fork == false
+    if: contains(github.event.pull_request.labels.*.name, 'vault-only')
```
[the remark this answers is not in the document]
Comment (agorokh) [reply .github/workflows/vault-automerge.yml:20]: Same comment as the line-12 thread — the `unlabeled` trigger removal is intentional per the upstream template. Not changing here; will file a follow-up against `agorokh/template-repo` so all child repos get any restoration together.

File: .github/workflows/post-merge-notify.yml [hunk 7]
```diff
@@ -7,21 +7,20 @@ on:
...
     steps:
-      - uses: actions/checkout@v4
-        with:
-          ref: ${{ github.event.pull_request.merge_commit_sha }}
-      - uses: actions/setup-python@v5
+      - uses: actions/checkout@v6
+
+      - uses: actions/setup-python@v6
         with:
           python-version: "3.11"
```
[the remark this answers is not in the document]
Comment (agorokh) [reply .github/workflows/post-merge-notify.yml:22]: The `ref: merge_commit_sha` removal is from the upstream template and the classifier script (`scripts/post_merge_classify.py`) reads PR diff via `gh api` rather than the local checkout, so it does not depend on the checked-out ref. That said, pinning the checkout to `merge_commit_sha` is defensive (protects against future script changes), so I will file a follow-up against `agorokh/template-repo` to restore it. Not changing in this sync PR to avoid template drift.

File: .claude/agents/issue-driven-coding-orchestrator.md [hunk 8]
```diff
@@ -54,6 +54,7 @@ memory: project
 10. **Mark the PR ready for review:** use pull request number **P** (from the PR you opened; it can differ from issue **N**): `gh pr ready <P> --repo <owner/repo>` (exits draft state so reviewers and bots are notified).
 11. **Wait for async bots before handoff:** follow **§ Mandatory wait after each push** in `.claude/agents/pr-resolution-follow-up.md`—use the same **`sleep 600`** cooldown after `gh pr ready` as after a `git push` (that section owns the normative wait; do not skip it).
 12. Hand off to **PR resolution** using the same pull request number **P** as in step 10: in Claude Code invoke **`Task(subagent_type="pr-resolution-follow-up", …)`**; in Cursor use **`Task(subagent_type="generalPurpose", …)`** with the **pr-resolution-follow-up** checklist from `.claude/agents/pr-resolution-follow-up.md`, or execute that file’s steps inline until CI is green and bot threads are addressed.
+13. **Fresh file reads before edits:** before substantive **Edit**/**Write** on paths you will change, load current contents for those targets (in Claude Code use **`query_file_patterns`** with those paths; in other tools use an equivalent bulk read) so patches apply to the latest revisions.
```
[the remark this answers is not in the document]
Comment (agorokh) [reply .claude/agents/issue-driven-coding-orchestrator.md:57]: Acknowledged. The ordering is a real issue (step 13 is after handoff, so agents reach it too late), but the file comes verbatim from `template-repo@061d9ab` and re-ordering here would create drift. Will file a follow-up against `agorokh/template-repo` to move the pre-edit guidance earlier in the sequence. Not blocking this sync PR.

File: scripts/hook_protect_main_impl.py [hunk 9]
```diff
@@ -0,0 +1,621 @@
...
+    git_head_argv = args[0:j] + ["symbolic-ref", "--short", "HEAD"]
+    try:
+        r = subprocess.run(
+            git_head_argv,
+            capture_output=True,
+            text=True,
+            timeout=_GIT_HEAD_TIMEOUT_SEC,
+            env=shell_env,
+        )
+    except (FileNotFoundError, OSError) as exc:
```
[the remark this answers is not in the document]
Comment (agorokh) [reply scripts/hook_protect_main_impl.py:411]: Valid robustness suggestion, but the file is unchanged from upstream `template-repo@061d9ab` in this sync PR. Deferring to an upstream issue so the fix lands once and propagates to every child repo rather than creating template drift here.

File: scripts/post_merge_sync.sh [hunk 10]
```diff
@@ -29,138 +49,317 @@ if [[ -z "$PR" ]] || ! [[ "$PR" =~ ^[0-9]+$ ]]; then
...
-  if ! command -v gh >/dev/null 2>&1; then
-    echo "error: gh CLI not found" >&2
-    exit 20
+STASH_LABEL="post-merge-pr${PR}-wip"
+SAVED_STASH=""
+
+stash_wip_if_dirty() {
+  if ! git diff --quiet || ! git diff --cached --quiet || [[ -n "$(git ls-files --others --exclude-standard)" ]]; then
+    echo "Working tree dirty; stashing as '${STASH_LABEL}'..."
+    if git stash push --include-untracked --message "$STASH_LABEL" >/dev/null 2>&1; then
```
[the remark this answers is not in the document]
Comment (agorokh) [reply scripts/post_merge_sync.sh:74]: Reasonable suggestion on stash label uniqueness, but this file is unchanged from upstream `template-repo@061d9ab`. Deferring to a follow-up issue against the template so the fix is authoritative and propagates uniformly.

File: tools/pr_pain/pain_score.py [hunk 11]
```diff
@@ -0,0 +1,565 @@
...
+            # not an array. Append as-is and let the caller handle it.
+            out.append(page)
+    return out
+
+
+_COMMITS_REST_CAP = 250  # GitHub REST `pulls/{pr}/commits` hard limit per PR.
+
+
+def _fetch_commits(repo: str, pr: int, gh: str) -> list[dict[str, Any]]:
+    """Fetch PR commits via REST.
```
[the remark this answers is not in the document]
Comment (agorokh) [reply tools/pr_pain/pain_score.py:280]: Accepted — REST caps at 250 commits per PR and the GraphQL switch would genuinely improve `fix_commits` / `commits_after_first_review` accuracy for large PRs. Same rationale as the other review threads on this PR: the file is unchanged from upstream and this is a template sync, so the fix belongs in `agorokh/template-repo` where it propagates to every child. Will file a follow-up issue.

File: .claude/agents/post-merge-steward.md [hunk 12]
```diff
@@ -22,35 +22,86 @@ memory: project
...
+| 2    | Bad usage | Stop; report missing arg |
+| 10   | Git conflict (ff-pull, stash pop, vault commit) | Stop; print stash ref + manual recovery hint; do not retry |
+| 11   | Branch protection rejected vault push | Stop; print PR URL; ask human to merge or grant bot bypass |
+| 12   | PR state unexpected / `gh pr create` failed | Stop; report state; do not retry merge |
+| 13   | Linked issue still OPEN | Surface in handoff; do not auto-close |
+| 20   | Infrastructure error (`gh`/`git` missing, auth, network) | Stop; report; do not retry |
+| 30   | Vault scope violation (non-vault tracked changes) | Stop; tell human to commit/stash other work first |
 
-## Exit codes
+**The agent must not improvise on any non-zero exit.** Print the code, the suggested action, and stop. No `SKIP=...`, no `--no-verify` outside what the script already does, no manual `git push origin main`, no asking the human to push.
```
[the remark this answers is not in the document]
Comment (agorokh) [reply .claude/agents/post-merge-steward.md:69]: Noted. Line 69 is the "no improvisation on non-zero exit" rule; the perceived conflict with the "surface exit 13 in handoff" rule (line 65) is because 13 is an informational exit (linked issue still open) rather than a failure: Phase A completed, Phase C can still run, and the agent surfaces the issue state in the handoff. I will clarify the wording upstream in `agorokh/template-repo` rather than patching the synced file here.

**Comment by agorokh** [comment]
@coderabbitai review

**Comment by agorokh** [comment]
C:/Program Files/Git/gemini review

**Comment by agorokh** [comment]
C:/Program Files/Git/review

**Comment by agorokh** [comment]
@sourcery-ai review

**Comment by SourceryAI** [comment]
Sorry, we are unable to review this pull request

The GitHub API does not allow us to fetch diffs exceeding 20000 lines

**Comment by agorokh** [comment]
@coderabbitai review

**Comment by agorokh** [comment]
C:/Program Files/Git/gemini review

**Comment by agorokh** [comment]
C:/Program Files/Git/review

**Comment by agorokh** [comment]
@sourcery-ai review

**Comment by agorokh** [comment]
## Review round 2 disposition

All unresolved review threads on this sync PR fall into two buckets:

**Fixed in this PR**
- `tools/process_miner/github_client.py:327` (Copilot) — tightened `base64.b64decode` to `validate=True` so the existing exception handler actually triggers on non-base64 payloads. Fixed in 53bf74f.

**Deferred upstream — tracked in agorokh/template-repo#97**
Every other unresolved thread sits on a file that came in **unchanged** from `template-repo@061d9ab`. Fixing in this child repo would cause template drift across every spoke on the next sync. The upstream tracker (agorokh/template-repo#97) lists all 13 items with severity, reviewer attribution, and links back to the source threads here.

Two of those items are **P1** (from codex) on `scripts/post_merge_sync.sh` and deserve callouts:
- L125: re-check PR merge state after `gh pr merge` before declaring success
- L170: scope the `git branch -D` loop so it can't wipe unrelated in-progress branches

Resolving the corresponding threads on this PR now that there's an upstream paper trail. The fix lands here via the next template sync.

---
Author: resolution executed by pr-resolution-follow-up per `.claude/agents/pr-resolution-follow-up.md`.

**Comment by SourceryAI** [comment]
Sorry, we are unable to review this pull request

The GitHub API does not allow us to fetch diffs exceeding 20000 lines

**Comment by agorokh** [comment]
## Round 3 (53bf74f) update

Copilot's re-review on the latest SHA surfaced 4 additional threads. All land on template-unchanged files and follow the same deferral policy as the batch above. Updated agorokh/template-repo#97 with:

- `.github/workflows/post-merge-notify.yml:26` — fork guard restoration (same concern as the cursor thread on line 14).
- `scripts/hook_bash_pre_tool.sh:15` — orchestrator fails closed when dependent scripts or `CLAUDE_PROJECT_DIR` are missing.
- `scripts/pre_commit_check_test_artifacts.py:42` — `check_file()` doesn't handle `OSError` from `read_text()`.

Threads resolved; fix lands here via the next template sync.

**Comment by agorokh** [comment]
## Round 4 (codex) update

codex filed 3 more threads post-Bugbot-completion:

- **P1** `scripts/check_vault_follow_up.sh:29` — real unstaged-path guard regression: `scripts/hook_bash_pre_tool.sh` invokes the check *before* `git commit`, so the new early-exit on empty `--cached` diff lets unstaged tracked edits through.
- **P2** `scripts/post_merge_sync.sh:123` — auto-merge phase doesn't validate the PR targets `main`; bad PR number merges a non-main PR.
- **P2** `.github/workflows/vault-automerge.yml:12` — without the `unlabeled` event, there's no revocation path once `vault-only` enabled auto-merge.

All three on template-unchanged files; added to agorokh/template-repo#97 with codex's rationale preserved. Fix lands here via next template sync.
